azure-iot-device/tests/common/pipeline/pipeline_ops_test.py

```diff
@@ -0,0 +1,461 @@
+            # ONLY the provided kwargs are used - because there ARE additional kwargs added.
+            for kwarg in worker_op_kwargs:
+                assert mock_type.call_args[1][kwarg] == worker_op_kwargs[kwarg]
+
+        @pytest.mark.it(
+            "Adds a secondary callback to the worker operation after instantiation, if 'callback' is included in the provided **kwargs parameters"
+        )
+        def test_adds_callback_to_worker_op(self, mocker, op, worker_op_kwargs):
+            mock_instance = mocker.MagicMock()
+            mock_type = mocker.MagicMock(return_value=mock_instance)
+            mock_type.__name__ = "mock type"  # this is needed for log statements
+            worker_op_kwargs["callback"] = mocker.MagicMock()
+
+            worker_op = op.spawn_worker_op(mock_type, **worker_op_kwargs)
+
+            assert worker_op is mock_instance
+            assert mock_type.call_count == 1
+
+            # The callback used for instantiating the worker operation is NOT the callback provided in **kwargs
+            assert mock_type.call_args[1]["callback"] is not worker_op_kwargs["callback"]
+
+            # The callback provided in **kwargs is applied after instantiation
+            assert mock_instance.add_callback.call_count == 1
+            assert mock_instance.add_callback.call_args == mocker.call(worker_op_kwargs["callback"])
+
+        @pytest.mark.it(
+            "Raises TypeError if the provided **kwargs parameters do not match the constructor for the class provided in the 'worker_op_type' parameter"
+        )
+        def test_incorrect_kwargs(self, mocker, op, worker_op_type, worker_op_kwargs):
+            worker_op_kwargs["invalid_kwarg"] = "some value"
+
+            with pytest.raises(TypeError):
+                op.spawn_worker_op(worker_op_type, **worker_op_kwargs)
+
+        @pytest.mark.it(
+            "Returns a worker operation, which, when completed, completes the operation that spawned it"
+        )
+        @pytest.mark.parametrize(
+            "use_error", [pytest.param(False, id="No Error"), pytest.param(True, id="With Error")]
+        )
+        def test_worker_op_completes_original_op(
+            self, mocker, use_error, arbitrary_exception, op, worker_op_type, worker_op_kwargs
+        ):
+            if use_error:
+                error = arbitrary_exception
+            else:
+                error = None
+
+            worker_op = op.spawn_worker_op(worker_op_type, **worker_op_kwargs)
+            assert op.complete.call_count == 0
+
+            worker_op.complete(error=error)
+
+            assert op.complete.call_count == 1
+            assert op.complete.call_args == mocker.call(error=error)
+
+        @pytest.mark.it(
+            "Returns a worker operation, which, when completed, triggers the 'callback' optionally provided in the **kwargs parameter, prior to completing the operation that spawned it"
```

Comment:
> prior to completing the operation that spawned it [](start = 134, length = 50)

This doesn't actually test the order that these callbacks are called.

Reply:
It actually does, due to the assertion in the callback. However, there were some bugs in this test anyway, which I have fixed up, as well as adding some better clarity.

---
In reply to: [346495978](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346495978) [](ancestors = 346495978)